nova/virt/vmwareapi/volumeops.py

```diff
@@ -590,3 +591,74 @@ def attach_root_volume(self, connection_info, instance,
         # we don't do that. This comment should help us detect upstream changes
         # to the function as merging should fail.
         self.attach_volume(connection_info, instance, adapter_type)
+
+    def fixup_shadow_vms(self, instance, shadow_vms):
+        # This should sensibly moved out to cinder
+        vm_ref = vm_util.get_vm_ref(self._session, instance)
+        for device in vm_util.get_hardware_devices(self._session, vm_ref):
+            class_name = device.__class__.__name__
+
+            if class_name != "VirtualDisk":
+                continue
+
+            try:
+                current_device_path = device.backing.fileName
+                volume = shadow_vms[device.backing.uuid]["volume"]
+                volume_ref = vutil.get_moref(volume, "VirtualMachine")
+                original_device = self._get_vmdk_base_volume_device(volume_ref)
+                if original_device:
+                    original_device_path = original_device.backing.fileName
+                    if original_device_path == current_device_path:
+                        # Already attached
+                        continue
+                    else:
+                        # That should not happen
+                        LOG.warning("Shadow-vm %s has already disk"
+                            " attached at {} replacing it with {}",
+                            volume, original_device_path, current_device_path
+                            )
+                        self.detach_disk_from_vm(self, volume_ref, instance,
+                            original_device, destroy_disk=True)
+
+                disk_type = vm_util._get_device_disk_type(device)
+                self.attach_disk_to_vm(volume_ref,
+                    instance,
+                    constants.DEFAULT_ADAPTER_TYPE,
+                    disk_type,
+                    current_device_path
+                    )
+            except KeyError:
+                # Unknown volume, so we ignore it
+                continue
+            except Exception:
+                LOG.exception("Failed to attach volume {}".format(device.key))
+
+    def delete_shadow_vms(self, block_device_info, instance=None):
+        # We need to delete the migrated shadow vms
+        # (until we implement it in cinder)
+        block_device_mapping = driver.block_device_info_get_mapping(
+            block_device_info)
+
+        if not block_device_mapping:
+            return
+
+        session = self._session
+        deleted = []
+        for disk in block_device_mapping:
+            connection_info = disk["connection_info"]
+            try:
+                data = connection_info["data"]
+                volume_ref = self._get_volume_ref(data["volume"])
```

Comment:
Not quite, the function is called in  `post_live_migration` and `rollback_live_migration_at_destination`.

In `rollback_live_migration_at_destination`, the `block_device_info` is the "current" state of the failed migration. In other words, it is pointing to shadow-vms on the destination vcenter, with no data attached.
We better get rid of them as not to leak them.


From `post_live_migration` (which is called on the source) the function gets passed the variable `source_bdms`, which stores the block-device-mapping at the source prior to the live-migration.
Since we have completed the live-migration successfully, the data (also) resides in the target vcenter attached to the respective shadow-vms. We can now delete the shadow-vms on the source and the associated data.

Reply:
Ah. This comes from somewhere else. I thought we're calling `_get_volume_mappings()` here. Then it makes sense.

Reply:
Note though, that the caller already filters out all volumes, where `connection_info` is empty

Reply:
But we also fail the live-migration if we encounter a volume without `connection_info` and this is only called later, so that should not be too problematic, right?

Reply:
In the code I have not yet pushed yes.

In the code as it is now, we only fail the migration on a missing `connection_info` for the attachments on the new compute-host, but `post_live_migration` works on the attachments for the old-compute host, and there the manager would simply have filtered out those volumes.

Reply:
Now I check the connection_info before the migration.